File: ios/GhostChat/Core/Files/ChunkTimeoutTracker.swift
import Foundation

/// Per-file chunk timeout tracker with bounded retries.
///
/// Semantics:
/// - `arm(fileId:)` — start watching an incoming file. Initialises the retry
///   counter to 0 and schedules a deadline `timeout` seconds in the future.
/// - `progressed(fileId:)` — a chunk arrived. Reset the deadline AND reset
///   retry counter to 0 (the transfer is making progress).
/// - `cancel(fileId:)` — transfer completed or was aborted by the peer.
///
/// When the deadline fires without a `progressed`:
///   1. If `retries < maxRetries`, bump `retries`, call `onTimeout(fileId)` so
///      the caller can request a retransmit, and arm a fresh deadline.
///   2. Otherwise call `onAbort(fileId)` and stop watching the file.
///
/// Mirror of Android `ChunkTimeoutTracker` — identical state machine + numbers.
final class ChunkTimeoutTracker {

    /// Seconds of silence that trigger a timeout. Spec: 30 seconds.
    static let defaultTimeout: TimeInterval = 30.0
    /// How many back-to-back timeouts we tolerate before aborting the transfer.
    /// Spec: 3 retries then abort.
    static let defaultMaxRetries: Int = 3

    private struct Entry {
        var deadlineTask: Task<Void, Never>
        var retries: Int
    }

    private let timeout: TimeInterval
    private let maxRetries: Int
    private let lock = NSLock()
    private var entries: [String: Entry] = [:]

    /// Called when `timeout` seconds elapsed with no `progressed` and we still
    /// have retry budget. The caller typically requests a retransmit of missing
    /// chunks. Callback is invoked off the main actor — dispatch as needed.
    var onTimeout: ((String) -> Void)?

    /// Called after the last retry's deadline fires without progress. The
    /// tracker has already stopped watching this fileId by the time the
    /// callback runs.
    var onAbort: ((String) -> Void)?

    init(
        timeout: TimeInterval = ChunkTimeoutTracker.defaultTimeout,
        maxRetries: Int = ChunkTimeoutTracker.defaultMaxRetries
    ) {
        self.timeout = timeout
        self.maxRetries = maxRetries
    }

    /// Begin tracking `fileId`. If already tracking it, the prior timer is
    /// cancelled and retry counter reset.
    func arm(fileId: String) {
        lock.lock(); defer { lock.unlock() }
        entries[fileId]?.deadlineTask.cancel()
        entries[fileId] = Entry(deadlineTask: schedule(fileId), retries: 0)
    }

    /// A chunk arrived. Resets the deadline AND the retry counter.
    func progressed(fileId: String) {
        lock.lock(); defer { lock.unlock() }
        guard entries[fileId] != nil else { return }
        entries[fileId]?.deadlineTask.cancel()
        entries[fileId] = Entry(deadlineTask: schedule(fileId), retries: 0)
    }

    /// Transfer finished (success or failure). No further timers.
    func cancel(fileId: String) {
        lock.lock(); defer { lock.unlock() }
        entries[fileId]?.deadlineTask.cancel()
        entries.removeValue(forKey: fileId)
    }

    /// Cancel every armed timer. Called on full session teardown.
    func cancelAll() {
        lock.lock(); defer { lock.unlock() }
        for (_, entry) in entries { entry.deadlineTask.cancel() }
        entries.removeAll()
    }

    /// Number of files currently being tracked. Test-only / observability.
    var armedCount: Int {
        lock.lock(); defer { lock.unlock() }
        return entries.count
    }

    // MARK: - Private

    private func schedule(_ fileId: String) -> Task<Void, Never> {
        let t = timeout
        return Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(t * 1_000_000_000))
            if Task.isCancelled { return }
            self?.fireTimeout(fileId: fileId)
        }
    }

    private func fireTimeout(fileId: String) {
        lock.lock()
        guard var entry = entries[fileId] else { lock.unlock(); return }
        let nextRetries = entry.retries + 1
        if nextRetries > maxRetries {
            entries.removeValue(forKey: fileId)
            lock.unlock()
            onAbort?(fileId)
            return
        }
        entry.retries = nextRetries
        entry.deadlineTask = schedule(fileId)
        entries[fileId] = entry
        lock.unlock()
        onTimeout?(fileId)
    }
}

File: android/app/src/main/java/com/kordar/ghostchat/core/files/ChunkTimeoutTracker.kt
package com.kordar.ghostchat.core.files

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.util.concurrent.ConcurrentHashMap

/**
 * Per-file chunk timeout tracker with bounded retries.
 *
 * Semantics (mirror iOS `ChunkTimeoutTracker`):
 * - [arm] — start watching an incoming file. Initialises retries to 0 and
 *   schedules a deadline [timeoutMs] ms in the future.
 * - [progressed] — a chunk arrived. Reset the deadline AND reset the retry
 *   counter to 0 (the transfer is making progress).
 * - [cancel] — transfer completed or aborted by peer.
 *
 * When the deadline fires without a [progressed]:
 *   1. If `retries < maxRetries`, bump retries, call [onTimeout] so the caller
 *      can request a retransmit, and arm a fresh deadline.
 *   2. Otherwise call [onAbort] and stop watching the file.
 */
class ChunkTimeoutTracker(
    private val timeoutMs: Long = DEFAULT_TIMEOUT_MS,
    private val maxRetries: Int = DEFAULT_MAX_RETRIES,
    private val scope: CoroutineScope = CoroutineScope(
        SupervisorJob() + Dispatchers.Default
    )
) {
    companion object {
        /** Spec: 30 seconds of silence triggers a timeout. */
        const val DEFAULT_TIMEOUT_MS: Long = 30_000L
        /** Spec: 3 back-to-back timeouts then abort. */
        const val DEFAULT_MAX_RETRIES: Int = 3
    }

    private data class Entry(val job: Job, val retries: Int)

    private val entries = ConcurrentHashMap<String, Entry>()

    /** Fires when [timeoutMs] ms elapsed with no [progressed] and we still
     *  have retry budget. Caller should request a retransmit of missing chunks. */
    var onTimeout: ((String) -> Unit)? = null

    /** Fires after the last retry's deadline fires without progress. The
     *  tracker has already stopped watching this fileId. */
    var onAbort: ((String) -> Unit)? = null

    fun arm(fileId: String) {
        entries[fileId]?.job?.cancel()
        entries[fileId] = Entry(schedule(fileId), 0)
    }

    fun progressed(fileId: String) {
        val current = entries[fileId] ?: return
        current.job.cancel()
        entries[fileId] = Entry(schedule(fileId), 0)
    }

    fun cancel(fileId: String) {
        entries.remove(fileId)?.job?.cancel()
    }

    /** Cancel every armed timer. Called on full session teardown. */
    fun cancelAll() {
        val snapshot = entries.toMap()
        entries.clear()
        for ((_, entry) in snapshot) entry.job.cancel()
    }

    /** Number of files currently being tracked. Test-only / observability. */
    val armedCount: Int get() = entries.size

    // region Private

    private fun schedule(fileId: String): Job = scope.launch {
        delay(timeoutMs)
        // If the coroutine was cancelled while sleeping, `delay` already threw
        // CancellationException and we won't reach here. Explicit check is for
        // defence-in-depth in case cancellation lands between the delay returning
        // and the body resuming.
        if (!isActive) return@launch
        fireTimeout(fileId)
    }

    private fun fireTimeout(fileId: String) {
        val current = entries[fileId] ?: return
        val nextRetries = current.retries + 1
        if (nextRetries > maxRetries) {
            entries.remove(fileId)
            onAbort?.invoke(fileId)
            return
        }
        entries[fileId] = Entry(schedule(fileId), nextRetries)
        onTimeout?.invoke(fileId)
    }

    // endregion
}
